core/trino-main/src/main/java/io/trino/sql/planner/EqualityInference.java

```diff
@@ -211,7 +212,7 @@ public static EqualityInference newInstance(Metadata metadata, Collection<Expres
         Collection<Set<Expression>> equivalentClasses = equalities.getEquivalentClasses();
 
         // Map every expression to the set of equivalent expressions
-        Map<Expression, Set<Expression>> byExpression = new HashMap<>();
+        Map<Expression, Set<Expression>> byExpression = new LinkedHashMap<>();
```

Comment:
Why?

Reply:
We don't want plan generation to be non-deterministic. Other sets in this class are already deterministic

Reply:
There is enumeration below:
```
        // For every non-derived expression, extract the sub-expressions and see if they can be rewritten as other expressions. If so,
        // use this new information to update the known equalities.
        Set<Expression> derivedExpressions = new LinkedHashSet<>();
        for (Expression expression : byExpression.keySet()) {
```

if you have `cos(x)=0, x=y, cos(y)=0` then visit order might matter to what will be considered a `derivedExpressions` expression.

Reply:
Can you add a test that would fail (sporadically) due to this?

Reply:
I was thinking about something like:
```
    @Test(invocationCount = 100)
    public void testDerivedExpressionDeterminism()
    {
        EqualityInference inference = EqualityInference.newInstance(
                metadata,
                equals(nameReference("a"), add("b", "z")),
                equals(nameReference("b"), nameReference("d")),
                equals(nameReference("a"), add("d", "z")));
        // generated equalities should be deterministic
        assertEquals(
                inference.generateEqualitiesPartitionedBy(symbols("a", "b", "z", "d")).getScopeEqualities(),
                ImmutableList.of(equals(nameReference("a"), add("b", "z")), equals(nameReference("b"), nameReference("d"))));
    }
```
but `generateEqualitiesPartitionedBy` doesn't filter derived expressions when rewriting so it kind of work.

It might work deterministically with current code, but I'm not 100% sure. Hence I would land it anyway